### PR TITLE
DID Doc: Fix Marshaling

### LINKED ISSUE #2149: did.Doc should implement json.Marshaler
Users typically expect `json.Marshal(didDoc)` to produce a correctly formatted did document but that is not the case. Users are not typically aware of `did.Doc.JSONBytes()`. 

We should implement `MarshalJSON()` to avoid surprises.

### PR DESCRIPTION
**Pull request by gnarula**
Implemented json.Marshaler and json.Unmarshaler. Also changed
DIDCommMsgMap to use did.ParseDocument.

Closes #2149

### CODE REVIEW

File: pkg/didcomm/common/service/did_comm_msg.go [hunk 1]
```diff
@@ -203,6 +204,14 @@ func (m DIDCommMsgMap) Decode(v interface{}) error {
 				return base64.StdEncoding.DecodeString(v.(string))
 			}
 
+			if rt2 == reflect.TypeOf(did.Doc{}) {
+				didDoc, err := json.Marshal(v)
+				if err != nil {
+					return nil, err
```
Comment (llorllale) [inline pkg/didcomm/common/service/did_comm_msg.go:210]: please wrap this error

**Comment by llorllale** [comment]
@gnarula seems like you've uncovered bugs in the `did:peer` tests

**Comment by pfeairheller** [comment]
I hit this problem trying to parse messages from other frameworks.  What is needed to help move this along?  (yes, I'm volunteering).

**Comment by llorllale** [comment]
@gnarula @pfeairheller apologies for the long delay. I haven't had time to look too much into this, but, after glancing at the tests, my impression is the immediate problem is the setup: `TestComputeDID` expects a method-specific-ID computed from an peer DID doc that was marshalled incorrectly, presumably using `json.Marshal()`. This PR exposed that bug.

A quick way to unblock this PR would be to:

* fix the linter errors
* update the ID in did_test.go#L24 such that it matches the doc under test (L101-126).
* squash comits

Beyond that, in a followup, we should revisit this implementation and ensure alignment with the `did:peer` spec (we should coordinate with the Aries folks) and make sure the tests are being done right.

**Comment by llorllale** [comment]
@pfeairheller what's the source for those `did:peer` DIDs you're trying to parse?

**Comment by pfeairheller** [comment]
@llorllale interop testing with the IBM Verify Creds platform.  From looking at their logs, they are using the nodejs wrapper on indy-sdk.  And it was parsing a DIDDoc embedded inside a `connection` attribute in `didexchange` that errored.

**Comment by llorllale** [comment]
Hi @gnarula 

@pfeairheller was kind enough to submit PR #2340 which as merged.

Closing this PR.

**Pull request closed without merging**